Add navigation intro modal

joshuatf opened the pull request:
Fixes #6214 

Adds the video intro modal.

### Screenshots

<img width="697" alt="Screen Shot 2021-02-18 at 10 29 50 PM" src="https://user-images.githubusercontent.com/10561050/108455357-17554300-723c-11eb-9da3-563ca997029e.png">



### Detailed test instructions:

1. Enable the new navigation under WooCommerce -> Settings -> Advanced -> Features. (This will also require opting into tracking).
2. Visit the WooCommerce Admin homescreen.
3. Note the new modal.
4. Check that pagination works as expected and modal styling is as expected.
5. Dismiss the modal.
6. Refresh the page to verify the modal does not reappear.

1. On a new site, enable the navigation before visiting the homescreen.
2. Navigate to the homescreen.
3. Note the welcome modal is shown and the navigation intro modal is not shown.
4. Refresh the page and note the nav intro modal was dismissed and never shown.

<!-- 
Please add your test instructions to `/TESTING-INSTRUCTIONS.md`.
-->

<!--- Please add a Changelog note

Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance: to readme.txt under the "unreleased" list at the top of the changelog. If none exists, please add it. Also include PR number.

If changes pertain to a package, also update CHANGELOG.md in the package's folder in a similar manner.--->

joshuatf committed:
Add initial intro modal

joshuatf committed:
Hide modal on dismissal

joshuatf committed:
Add modal styling

joshuatf committed:
Add modal page content

joshuatf committed:
Update styles to match Figma designs

joshuatf committed:
Replace video with gifs

joshuatf committed:
Check if user is opted in to satisfy gdpr concerns

joshuatf committed:
Add testing instructions and changelog notes

elizaan36 commented:
looking good! Cool how you were able to remove the margin around the modal. A few fine-tuning things:

- Don't need the `Previous` link since people won't have much of a need to go back
- The text on the left hand side needs more padding-top so it's equal with the bottom, and probably more padding-right so it's vertically centered as well.
- I'm assuming the focus outline on the progress dots is something from Gutenberg but it's really distracting. Is that something we could remove?🙏

What do you think about the gifs? The resolution looks good to me but let me know if they need to be tweaked at all.

I went through the modal too fast and didn't double-check the copy but if it's the same as [this comment](https://github.com/woocommerce/woocommerce-admin/issues/6214#issuecomment-781427826) then we're good.

joshuatf committed:
Handle PR feedback

joshuatf commented:
Thanks for taking a look at this one, @elizaan36!

>Don't need the Previous link since people won't have much of a need to go back

Removed.

>The text on the left hand side needs more padding-top so it's equal with the bottom, and probably more padding-right so it's vertically centered as well.

It sounds like centering is the main concern, so I adjusted this so it's always vertically centered.  Had to make some small adjustments to padding.

With 40px here, we do have some cramming of the last title.  Here are all the slides with the updated style:

<img width="707" alt="Screen Shot 2021-02-19 at 8 47 06 AM" src="https://user-images.githubusercontent.com/10561050/108512699-9c1d7c80-728f-11eb-9187-7d91be3b193a.png">
<img width="687" alt="Screen Shot 2021-02-19 at 8 47 56 AM" src="https://user-images.githubusercontent.com/10561050/108512702-9cb61300-728f-11eb-8e68-50503932ae2d.png">
<img width="688" alt="Screen Shot 2021-02-19 at 8 47 59 AM" src="https://user-images.githubusercontent.com/10561050/108512704-9cb61300-728f-11eb-90b4-d2ef44f903e9.png">


>I'm assuming the focus outline on the progress dots is something from Gutenberg but it's really distracting. Is that something we could remove?🙏

It is from Gutenberg.  I think it may be an accessibility concern to remove it, but I'm not entirely sure.  @psealock can you weigh in here?

>What do you think about the gifs? The resolution looks good to me but let me know if they need to be tweaked at all.

I do notice a very minor amount of pixelation on a retina screen, but it's not bad.  I wouldn't be opposed to bumping up the dimensions of the images by 50% if it doesn't increase the asset size too much, but this is also nitpicky and I think the current resolution is also okay as is.

elizaan36 commented:
I updated the gifs to see if we can get a liiiiittle better resolution. [Here's the new folder](https://share.getcloudapp.com/llu964Zq)

> With 40px here, we do have some cramming of the last title.

Thanks for screenshots of the three pages. I really like the H2 style there so don't want to mess with that. I'm curious about the `max width: 85%` - maybe if we tweak that it wouldn't feel as crammed? Or using `$padding-large` all around may be the best way to go (to replace `$padding-largest`) as there's room to spare from the left and right edges. Otherwise I'm not too worried about it as it looks great overall! Perfectly centered.

> It is from Gutenberg. I think it may be an accessibility concern to remove it, but I'm not entirely sure. @psealock can you weigh in here?

The guide we're using on the Home screen modal doesn't show the accessibility dots, so at least we thought it was ok to remove there.😆

I'm excited about this modal! Quick question - this wouldn't have any chance of interfering with the existing welcome modal on the home screen, would it? If someone just created a store and enables the new nav before visiting the home screen maybe? I wonder if there's a check we can include to make sure there isn't a conflict there.

elizaan36 commented:
We had briefly talked about mobile in Slack as well. I suggested to hide the modal completely on mobile but the alternative is to wrap the gif under the text so that at screen sizes below 680px the modal changes to 340px width.

joshuatf commented:
>The guide we're using on the Home screen modal doesn't show the accessibility dots, so at least we thought it was ok to remove there.😆

Oh, the dots aren't for accessibility, but the outline for focus around them I believe is.  Do you want to remove the dots entirely?

>I'm curious about the max width: 85% - maybe if we tweak that it wouldn't feel as crammed? 

<img width="688" alt="Screen Shot 2021-02-19 at 2 53 21 PM" src="https://user-images.githubusercontent.com/10561050/108554869-b4f35580-72c2-11eb-889e-3e71a7fd5a75.png">

This is to prevent the line break that occurs.  But if we update the padding here to `$gap-large` then all the text sits okay:

<img width="683" alt="Screen Shot 2021-02-19 at 2 54 50 PM" src="https://user-images.githubusercontent.com/10561050/108554849-b0c73800-72c2-11eb-880c-f2ca2cc630ae.png">
<img width="688" alt="Screen Shot 2021-02-19 at 2 54 45 PM" src="https://user-images.githubusercontent.com/10561050/108554851-b15fce80-72c2-11eb-8f7d-949e0da44082.png">
<img width="690" alt="Screen Shot 2021-02-19 at 2 54 41 PM" src="https://user-images.githubusercontent.com/10561050/108554852-b15fce80-72c2-11eb-9061-63d57a2b6132.png">

>this wouldn't have any chance of interfering with the existing welcome modal on the home screen, would it? 

That's a good question.  I think it does have the potential if navigation was already enabled on a store (like ecomm plan users).  If that occurs, the navigation modal lays on top of the welcome modal.  Both modals function correctly, but users have to go through two modals when opening the page for the first time.  If we want to update this, we can show after the welcome modal is dismissed.

joshuatf committed:
Handle responsiveness

joshuatf commented:
And here's the responsive version.  Note the image is slightly truncated at the bottom to prevent the need to scroll the modal to see the "Next" buttons, but I don't think it impacts the visuals on the image too much:

<img width="373" alt="Screen Shot 2021-02-19 at 3 18 47 PM" src="https://user-images.githubusercontent.com/10561050/108557159-25e83c80-72c6-11eb-8e63-c4fb04047c5a.png">

elizaan36 commented:
Responsive and `$gap-large` looks great!

> Oh, the dots aren't for accessibility, but the outline for focus around them I believe is. Do you want to remove the dots entirely?

Nope, I should have specified that the home screen guide modal has the navigation dots but no focus outline. Since we were able to remove the focus outline on those, could we remove them here as well.

>  If we want to update this, we can show after the welcome modal is dismissed.

That's a good solution, but we could get into a situation where one modal gets dismissed and another pops up immediately. Could we turn this modal off for new users where the navigation is enabled by default?

joshuatf commented:
>Nope, I should have specified that the home screen guide modal has the navigation dots but no focus outline. Since we were able to remove the focus outline on those, could we remove them here as well.

I think we should avoid doing this.  In general, elements with click handlers should be focusable and show an outline (or some kind of visual indication) that they are focused to remain accessible for keyboard-only users.  [Here's a good article explaining this in more detail](https://www.a11yproject.com/posts/2013-01-25-never-remove-css-outlines/).

If the outline was removed from the other modals, we should definitely look at adding them back in.  If long-term this impacts the design in a negative way, we can look at workarounds to hide outlines when mouse movement is detected, but this might be a large issue to tackle.

>Could we turn this modal off for new users where the navigation is enabled by default?

Unfortunately, we have no way of distinguishing between users where the nav is on by default vs when it has actually been toggled on.

Another couple options:
* Hold off on showing the modal until the next page refresh.
* If the welcome modal has not been shown, go ahead and dismiss the navigation modal.

The second option is going to produce results _closest to_ not showing the modal when the nav is on by default (since it's likely they'll have been to the homescreen prior to turning the nav on in settings).

elizaan36 commented:
> I think we should avoid doing this. In general, elements with click handlers should be focusable and show an outline (or some kind of visual indication) that they are focused to remain accessible for keyboard-only users.

Thanks for sharing that article. I'd agree with you if the dots were the primary method of navigating through the modal. Removing the focus outline from the dots shouldn't negatively impact accessibility because keyboard users should navigate via the primary "Next" or "Close" actions. 

The dots are a secondary element that provide additional information about which page you're on (and just so happen to be a navigational element but that's not the primary purpose in this case.)

> If the welcome modal has not been shown, go ahead and dismiss the navigation modal.

Yeah, I agree this option is the closest. Sounds like the best way to go.👍

joshuatf committed:
Dismiss modal when welcome modal is shown simultaneously

joshuatf commented:
>Removing the focus outline from the dots shouldn't negatively impact accessibility because keyboard users should navigate via the primary "Next" or "Close" actions.

That's not quite how it works.  Any focusable element on a page should provide a visual cue that it is focused so that keyboard-only users can identify where they are at all times.

As it currently stands, this means that keyboard users have no way of identifying where they are focused for at least 3 tabs.

We can:
* Remove the pagination entirely.
* Modify the the focus state outline stylistically.
* Try and remove the ability to focus pagination buttons so that we can also remove the outline.

>Yeah, I agree this option is the closest. Sounds like the best way to go.👍

Okay, updated to this option.  Will add additional testing instructions above.

elizaan36 commented:
I hear you. As I mentioned in Slack I'm glad you brought it up since it's not ideal to remove a focus outline on an interactive element. I get that it should be avoided and don't want to convey that I'm suggesting a less accessible experience. Here's an idea.

Regarding the welcome modal for the home screen, it must place the focus outline on the primary action by default. 

I don't know if it's possible but if the focus outline appears on the primary action by default we'd have less of a problem. It's not important for the keyboard reader to focus on the dots as they're a mostly informational element.

psealock reviewed (approved):
Nice Josh, this is working well and the code looks nice and clean. I just made a small suggestion about exporting an option name cleanup.

I tested the Welcome Modal + Nav Intro modal and its also working as expected.

As far as the focus outline around the dots, I DO see them on the welcome modal only when tabbing, which is the expected behaviour. I'm seeing the same thing on this modal as well. With that I think the need to remove them is addressed and I agree that we shouldn't remove these visual cues from Gutenberg components anyways.

elizaan36 commented:
sounds good to me. If we've got the same behavior as the other Gutenberg modals (focus outline on the primary action upon opening the modal) then that's perfect. All interactive elements should have a focus outline when tabbing.

joshuatf committed:
Update shared modal option constants

joshuatf commented on client/navigation/components/intro-modal/index.js:
```diff
@@ -0,0 +1,131 @@
+import './style.scss';
+
+const introModalOption = 'woocommerce_navigation_intro_modal_dismissed';
+const welcomeModalOption = 'woocommerce_task_list_welcome_modal_dismissed';
```
Good idea.  Updated here for these.  I see some others that could benefit from this going forward so just a side note that we may look at creating a shared constants file for the client.

joshuatf commented:
>sounds good to me. If we've got the same behavior as the other Gutenberg modals (focus outline on the primary action upon opening the modal) then that's perfect. All interactive elements should have a focus outline when tabbing.

Just so it's clear what's happening here, these modals do not receive focus to the primary action when opened.  They receive focus and show the outline when tabbed or clicked.

I can also confirm that the welcome modal focus still contains an outline so this won't need a follow-up to fix.

<img width="532" alt="Screen Shot 2021-02-22 at 1 33 50 PM" src="https://user-images.githubusercontent.com/10561050/108753385-ab602c80-7512-11eb-90f5-513b609e3b77.png">
